x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/grouping/CategorizeTests.java

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.esql.expression.function.grouping;
+
+import com.carrotsearch.randomizedtesting.annotations.Name;
+import com.carrotsearch.randomizedtesting.annotations.ParametersFactory;
+
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.xpack.esql.core.expression.Expression;
+import org.elasticsearch.xpack.esql.core.tree.Source;
+import org.elasticsearch.xpack.esql.core.type.DataType;
+import org.elasticsearch.xpack.esql.expression.function.AbstractScalarFunctionTestCase;
+import org.elasticsearch.xpack.esql.expression.function.TestCaseSupplier;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Supplier;
+
+import static org.hamcrest.Matchers.equalTo;
+
+public class CategorizeTests extends AbstractScalarFunctionTestCase {
+    public CategorizeTests(@Name("TestCase") Supplier<TestCaseSupplier.TestCase> testCaseSupplier) {
+        this.testCase = testCaseSupplier.get();
+    }
+
+    @ParametersFactory
+    public static Iterable<Object[]> parameters() {
+        List<TestCaseSupplier> suppliers = new ArrayList<>();
+        for (DataType dataType : List.of(DataType.KEYWORD, DataType.TEXT)) {
+            suppliers.add(
+                new TestCaseSupplier(
+                    "text with " + dataType.typeName(),
+                    List.of(dataType),
+                    () -> new TestCaseSupplier.TestCase(
+                        List.of(new TestCaseSupplier.TypedData(new BytesRef("blah blah blah"), dataType, "f")),
+                        "CategorizeInternalEvaluator[v=Attribute[channel=0]]",
+                        DataType.INTEGER,
+                        equalTo(0)
```

Comment:
Can we test other outputs? Potentially with randomized inputs, like in other function tests

Reply:
Is it possible to create testcases, where the function `CATEGORIZE` is evaluated sequentially on multiple inputs while keeping its state? Comparable to [the CSV test in this PR](https://github.com/elastic/elasticsearch/pull/112860/files#diff-75ee54559881f083236d7abc04df0b14344a8f624e3519c4de0e00b59cc4266d)?

Evaluating it on a single input doesn't give meaningful tests: the result always is `0` (meaning this single line of text is a category by itself, with ID 0)

Reply:
`AbstractScalarFunctionTestCase` isn't designed for that, sadly. I wonder if we should not use it and make something that extends `AbstractFunctionTestCase`. Or, well, something else? I'm not sure. This isn't a "scalar" function because of it's state.

Reply:
Indeed, that's something useful to consider.

However, at the moment `AbstractScalarFunctionTestCase` does provide a ton of useful stuff. The `...WithDefaultChecks` checks nulls, multivalues, wrong types, memory leaks, ... These tests helped me a lot fixing my bugs.